db/operations.go

```diff
@@ -0,0 +1,98 @@
+//(C) Copyright [2020] Hewlett Packard Enterprise Development LP
+//
+//Licensed under the Apache License, Version 2.0 (the "License"); you may
+//not use this file except in compliance with the License. You may obtain
+//a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//Unless required by applicable law or agreed to in writing, software
+//distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+//WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+//License for the specific language governing permissions and limitations
+// under the License.
+
+package db
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/go-redis/redis"
+)
+
+// Using below variables as part of errors will enabling errors.Is() function
+var (
+	// ErrorKeyAlreadyExist is for identifing already exist error
+	ErrorKeyAlreadyExist = errors.New("Key already exist in DB")
+	// ErrorKeyNotFound is for identifing not found error
+	ErrorKeyNotFound = errors.New("Key not Found in DB")
+)
+
+const (
+	// scanPaginationSize defines the size of DB keys to be scanned on single query
+	scanPaginationSize = 100
+)
+
+// Create will create a new entry in DB for the value with the given table and key
+func (c *Client) Create(table, key string, data interface{}) (err error) {
+	dataByte, err := json.Marshal(data)
+	if err != nil {
+		return fmt.Errorf("while marshalling data, got: %v", err)
+	}
+	ok, err := c.pool.SetNX(generateKey(table, key), string(dataByte), 0).Result()
+	switch {
+	case !ok:
+		return fmt.Errorf(
+			"%w: %s",
+			ErrorKeyAlreadyExist,
+			fmt.Sprintf("An entry with key %s is already present in table %s", key, table),
+		)
+	case err != nil:
+		return fmt.Errorf(
+			"Creating new entry for value %v in table %s with key %s failed: %v",
+			data, table, key, err,
+		)
+	default:
+		return nil
+	}
+}
+
+// GetAllKeys will collect all the keys of provided table
+func (c *Client) GetAllKeys(table string) ([]string, error) {
+	var allKeys []string
+	var cursor uint64
+	for {
+		keys, c, err := c.pool.Scan(cursor, generateKey(table, "*"), scanPaginationSize).Result()
```

Comment:
what does the * mean ?

Reply:
@Bharath-KKB This function retrieves all the keys associated with a particular table, which can be used while doing the GET 0n the collections.

Reply:
I did not ask what this is doing. I said why we are trying to retrieve all keys.

Reply:
@Bharath-KKB We are not retrieving all the keys, but the keys associated with a particular type, lets say "Fabric", so we are retrieving all fabric keys for fabric collection with Fabric:*.

Reply:
Addressed

Reply:
What are the table names for :
1) Fabric
2) Switches
3) Ports
4) Zones
5) Address Pools
6) Endpoints ?

Reply:
@Bharath-KKB table names will be added in the next PR. file path to table name constants is https://github.com/a-ajith/PluginCiscoACI/blob/187bd7a7f57f1290b9bdb57d8bb882badfaac9d9/capmodel/common.go#L26

Reply:
@Bharath-KKB added pattern as a parameter to the GetAllKeysFromTable function along with table name. So now we are able to get all keys associated with table_name:pattern*.